src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs

```diff
@@ -375,10 +375,9 @@ private SuggestedActionSet InlineActions(SuggestedActionSet actionSet)
                     // See https://github.com/dotnet/roslyn/issues/29589
                     const bool includeSuppressionFixes = true;
 
-                    var fixes = Task.Run(
+                    var fixes = this.ThreadingContext.JoinableTaskFactory.Run(
                         () => _owner._codeFixService.GetFixesAsync(
-                                document, range.Span.ToTextSpan(), includeSuppressionFixes, isBlocking: true, addOperationScope, cancellationToken),
-                        cancellationToken).WaitAndGetResult(cancellationToken);
+                                document, range.Span.ToTextSpan(), includeSuppressionFixes, isBlocking: true, addOperationScope, cancellationToken));
```

Comment:
We really need to make a decision on whether we're going to fully adopt JTF, or not. Until now, the requirements for code fixes have been clear: code fixes are asynchronous, but they are not _allowed_ to have a UI thread dependency. Changing this requirement without a policy that works for the entire code base is going to lead us down a path of whack-a-mole with deadlocks.

Reply:
My take on it is this:
1. we can adopt jtf judiciously here to alleviate the deadlock.
2. we can work with nuget to get a bg-safe API to use for legacy projects. (it appears we have one already for modern CPS projects).
3. when we get that api, we can move this back to being JTF free.

@jasonmalinowski do you have thoughts here?

Reply:
I think we should disable this feature for legacy projects. We should focus our efforts on modern projects.

Reply:
As for this workaround - I'd add a comment that this use of JTF is temporary workaround.

---
In reply to: [414843809](https://github.com/dotnet/roslyn/pull/43630#discussion_r414843809) [](ancestors = 414843809)

Reply:
The risk I see with your steps @CyrusNajmabadi is that in the window between 1 and 3 anybody could write a code fix that uses JTF now works, and then gets broken when we do 3. And that window in time may not be small -- I assume we'd ship at least one or two minor releases in meantime. In some sense we used the code we had here as a (partial) block so anybody trying to do the "wrong thing" would find out because their code doesn't work; if it suddenly does work they won't find out until it's perhaps too late and our mutual customers are getting hangs. I look at that very much as an opening of the barn door: we might be able to close it again, but the horse has already left either way.

Do we have confidence that this fix actually works? Given this is potentially waiting on tasks that have already kicked off or might be using AsyncLazy or something else that isn't JTF-aware in the implementation of ICodeFixService, is JTF able to let the right things through here? My concern here being an independent question of "is JTF wise?" but rather "is JTF actually going to work here given we have an existing body of code that in places is not following JTF rules?"

I don't know if JTF has support built in but we _could_ do something where we have a handshake between the code fix at hand and the Wait() we do on the UI thread to make a SwitchToMainThreadAsync()-like method that is able to run code despite being in that wait. Conceptually something like "we are blocking jumps to the UI thread, unless you are this specific jump with a specific password, and then we let you through." Or some custom way to control exactly what is "let through". If it doesn't have support we _can_ implement a very constrained system that might meet exactly our needs here (and not try to solve the general problem) which at least means we can barn-door-opening in the general case. In some sense this means we're reinventing JTF but rather than making a general-purpose thing the specific purpose thing can be more precisely targeted. Is this evil? maybe, but it might be an alternative solution worth looking at if we haven't already.

More generally, out-of-proc is on the horizon here; I don't know if there we'd start running everything OOP except for things marked as "I need to be in-proc", and if those run with JTF it's already clear they're doing evil and the evilness is already called out.